yesod-core/Yesod/Core/FormHelpers.hs

```diff
@@ -0,0 +1,79 @@
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Yesod.Core.FormHelpers
+    (
+      HttpMethod(..),
+      FormStyle,
+      httpLink,
+      buttonLink,
+      link
+    ) where
+
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.Trans.Control (MonadBaseControl (..))
+import Yesod.Routes.Class
+import Control.Monad.Trans.Resource (MonadResource (..), MonadThrow (..))
+import Yesod.Core.Widget
+import Yesod.Core.Handler
+import Text.Blaze (ToMarkup)
+import Data.String (IsString)
+import Text.Lucius
+
+data HttpMethod = HttpGET | HttpPOST | HttpPATCH | HttpUPDATE | HttpDELETE
```

Comment:
I was attempting to add type security to the HTTP methods to force it to match up with one of the routes. This initially started when I was trying to use the Yesod generated routes for a form action to deduce what the correct HTTP action would be, but I couldn't figure out a good way to do that. I can remove this and just use text. 

This also leads me to another thought: What about making an addition to the generated routes that pairs together the action and the method so you can query which method someone expected in forms and such and make sure they're always hitting the correct method and the correct path? The reason I started this change was because my GET request was being made, but failing on the server because the end route was actually a POST.